ts/nni_manager/training_service/local_v3/local.ts

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Logger, getLogger } from 'common/log';
+import type { LocalConfig, TrainingServiceConfig } from 'common/experimentConfig';
+import type { EnvironmentInfo, Metric, Parameter, TrainingServiceV3 } from 'common/training_service_v3';
+import { TrialKeeper } from 'common/trial_keeper/keeper';
+
+export class LocalTrainingServiceV3 implements TrainingServiceV3 {
+    private config: LocalConfig;
+    private env: EnvironmentInfo;
+    private log: Logger;
+    private trialKeeper: TrialKeeper;
+
+    constructor(trainingServiceId: string, config: TrainingServiceConfig) {
+        this.log = getLogger(`LocalV3.${trainingServiceId}`);
+        this.log.debug('Training sevice config:', config);
+
+        this.config = config as LocalConfig;
+        this.env = { id: `${trainingServiceId}-env` };
+        this.trialKeeper = new TrialKeeper(this.env.id, 'local', Boolean(config.trialGpuNumber));
+    }
+
+    public async init(): Promise<void> {
+        return;
+    }
+
+    public async start(): Promise<EnvironmentInfo[]> {
+        this.log.info('Start');
+        await this.trialKeeper.start();
+        return [ this.env ];
+    }
+
+    public async stop(): Promise<void> {
+        await this.trialKeeper.shutdown();
+        this.log.info('All trials stopped');
+    }
+
+    /**
+     *  Note:
+     *  The directory is not copied, so changes in code directory will affect new trials.
+     *  This is different from all other training services.
+     **/
+    public async uploadDirectory(directoryName: string, path: string): Promise<void> {
+        this.log.info(`Register directory ${directoryName} = ${path}`);
+        this.trialKeeper.registerDirectory(directoryName, path);
+    }
+
+    public async createTrial(_envId: string, trialCommand: string, directoryName: string, sequenceId?: number):
+            Promise<string | null> {
+
+        const trialId = uuid();
+
+        let gpuNumber = this.config.trialGpuNumber;
+        if (gpuNumber) {
+            gpuNumber /= this.config.maxTrialNumberPerGpu;
+        }
+
+        const opts: TrialKeeper.TrialOptions = {
+            id: trialId,
+            command: trialCommand,
+            codeDirectoryName: directoryName,
+            sequenceId,
+            gpuNumber,
+            gpuRestrictions: {
+                onlyUseIndices: this.config.gpuIndices,
+                rejectActive: !this.config.useActiveGpu,
+            },
+        };
+
+        const success = await this.trialKeeper.createTrial(opts);
```

Comment:
Do JS-guys prefer `.done().error()` or something like this (`if (success)`)?

Reply:
In non-local training services this is an RPC, so exception is not a choice. (promise rejection === async exception)